cmd/node-agent/main.go

```diff
@@ -0,0 +1,302 @@
+		},
+	)
+	var configMapController *framework.Controller
+	configMapStore, configMapController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(lo api.ListOptions) (runtime.Object, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).List(lo)
+			},
+			WatchFunc: func(lo api.ListOptions) (watch.Interface, error) {
+				return client.Core().ConfigMaps(api.NamespaceSystem).Watch(lo)
+			},
+		},
+		&v1.ConfigMap{},
+		30*time.Second,
+		framework.ResourceEventHandlerFuncs{},
+	)
+
+	go configMapController.Run(wait.NeverStop)
+	nodeController.Run(wait.NeverStop)
+}
+
+func newAPIClient() clientset.Interface {
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		glog.Fatal(err)
+	}
+	return clientset.NewForConfigOrDie(config)
+}
+
+func nodeUpdateCallback(_, newObj interface{}) {
+	glog.Infof("begin node update callback")
+
+	newNode, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("received unexpected type: %T (expected *v1.Node)", newObj)
+		return
+	}
+	onDiskConfig, err := parseKubeletEnvFile(kubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	if newNode.Annotations == nil {
+		newNode.Annotations = make(map[string]string)
+	}
+	rawconf, ok := newNode.Annotations[desiredConfigAnnotation]
+	if !ok {
+		if _, ok := newNode.Annotations[currentConfigAnnotation]; !ok {
+			setCurrentConfigAnnotation(newNode, onDiskConfig)
+		}
+		glog.Info("no configuration annotation found, ignoring")
+		return
+	}
+	var desiredConfig map[string]string
+	if err = json.Unmarshal([]byte(rawconf), &desiredConfig); err != nil {
+		glog.Error(err)
+		return
+	}
+	if err = validateConfig(desiredConfig); err != nil {
+		glog.Error(err)
+		return
+	}
+
+	cm, err := getConfigMap(desiredConfig[kubeletConfigKey])
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	if configHasChanged(onDiskConfig, desiredConfig, cm) {
+		if err = updateKubeletEnvFile(desiredConfig, onDiskConfig, cm); err != nil {
+			glog.Error(err)
+			return
+		}
+		if err = restartKubeletService(); err != nil {
+			glog.Error(err)
+			return
+		}
+	}
+	// Re-parse on disk config as it is soure of truth.
+	onDiskConfig, err = parseKubeletEnvFile(kubeletEnvPath)
+	if err != nil {
+		glog.Error(err)
+		return
+	}
+	updatedConfig := make(map[string]string)
+	updatedConfig[kubeletVersionKey] = onDiskConfig[kubeletVersionKey]
+	updatedConfig[kubeletFlagsKey] = onDiskConfig[kubeletFlagsKey]
+	updatedConfig[kubeletConfigKey] = desiredConfig[kubeletConfigKey]
+	// Always update annotation at end of sync loop.
+	if err := updateCurrentConfigNodeAnnotation(newNode, updatedConfig); err != nil {
+		glog.Error(err)
+	}
+}
+
+func setCurrentConfigAnnotation(node *v1.Node, onDiskConfig map[string]string) {
+	currentConfig := make(map[string]string)
+	currentConfig[kubeletVersionKey] = onDiskConfig[kubeletVersionKey]
+	currentConfig[kubeletFlagsKey] = onDiskConfig[kubeletFlagsKey]
+	currentConfig[kubeletConfigKey] = "intiial on-disk configuration"
```

Comment:
nit: "on-disk configuration" -- not necessarily initial, as it could change over time. Or if left as is: s/intiial/initial/